CF-468: add a profile to submit war to Veracode

shintasmith opened the pull request:
The above profile is to be executed from a Jenkins job:
https://jenkins.inova.dfw1.ci.rackspace.net/view/Feeds%20All/job/cloudfeeds-atomhopper-veracode-upload-scan/

Once the current scan is completed, I will run the Jenkins job.

ChandraAddala commented on feeds-atomhopper/pom.xml:
```diff
@@ -246,5 +246,87 @@
+
+                                        // workaround for the veracode-client-java not taking buildVersion as String
+                                        def buildVersion = version - '-SNAPSHOT'
+                                        buildVersion = buildVersion.replaceAll("\\.","") + '${build.number}'
```
I am not sure if I understand this correctly but 1.11.1 and 1.1.11 would result in the same buildVersion right?

shintasmith replied:
That is correct. I don't have a better idea on how to get a "build" number. Should I just generate a random one? If so, there is no way to tie to a specific version. 

Veracode API takes only integers :-( Otherwise, I would've just used the version string.

ChandraAddala replied:
If we assume at each portion of the build string will only go from 0-99, we can do 

<main>.<major>.<minor>

minor + major*100 + main * 100 * 100 

1.12.35 would be 11235. 
1.1.20  would be 10120

Something like that. What do you think?

shintasmith replied:
Actually, as I was trying to make changes for your suggestion, I noticed that I am appending the jenkins build number at the end. 
```
                                        def buildVersion = version - '-SNAPSHOT'
                                        buildVersion = buildVersion.replaceAll("\\.","") + '${build.number}'
```
That ```${build.number}``` is a Jenkins variable that gets passed in too the job. It can also be set from the command line using ```-Dbuild.number=111``` (if you are running mvn directly from your box). 

So I think we should be ok. I am not going to make changes for the the buildVersion calculation.

ChandraAddala replied:
I retract all my comments. Turns out I dont care so much for veracode after all.

shintasmith replied:
Dont merge this yet. Now that @ChandraAddala got me started, I'm tweaking a few things. The buildVersion being integer is a limitation of the Rackspace Veracode client wrapper API. I'm trying to get around it.

ChandraAddala commented on feeds-atomhopper/pom.xml:
```diff
@@ -246,5 +246,87 @@
+                                        import groovyx.net.http.HTTPBuilder
+                                        import static groovyx.net.http.Method.*
+
+                                        def client = new DefaultVeracodeApiClient("https://analysiscenter.veracode.com/api/", '${veracode.username}', '${veracode.password}', System.out)
```
Should we also make https://analysiscenter.veracode.com/api/ also as parameter we get from jenkins job?

shintasmith replied:
I'm not sure there's a big benefit of doing that. I would rather not have the person who's running the Jenkins job (which could be NebOps admins) to have to know the right URL to Veracode API.

ChandraAddala replied:
The benefit I was thinking was if the URL gets changed, we dont have to make code changes.

shintasmith replied:
I actually think not having to make code changes is a drawback. >.< Because changes in the Jenkins job is not version controlled. 

But ok. Let me see if I can add the parameter with default value. I have to change this in the existing Atom Hopper job as well, to be consistent.

ChandraAddala commented on feeds-atomhopper/pom.xml:
```diff
@@ -246,5 +246,87 @@
+
+                                        // workaround for the java client not making HTTP call to beginscan.do
+                                        print "Submitting /api/4.0/beginscan.do\n"
+                                        def http = new HTTPBuilder("https://analysiscenter.veracode.com/api/")
```
If we are not externalizing this value, can we alteast declare a constant which holds the value of the url?

shintasmith replied:
Yeah, that makes sense. I'll do that.

usnavi commented:
dgad

shintasmith commented:
so I can do whatever I want?

shintasmith commented:
@ChandraAddala, @usnavi, I updated the Veracode script. I am no longer using the Rackspace Veracode client to upload the artifacts. That jar file imposed the unnecessary limitation that buildVersion has to be an integer. We can now upload version like "1.11.0-1". 

I also changed the script to use new APIs that allow me to get rid of the sleep of 5 minutes waiting on pre-scan to complete before starting the actual scan. There is an auto_scan flag that you can send in. If set to true, the actual scan will start as soon as pre-scan is completed.

Please review. Once you bless this, I will copy to other git repos.

ChandraAddala commented:
Looks good.
